File: src/crypto/surrogate.test.ts
import { describe, it, expect } from "vitest";
import { containsLoneSurrogateEscape, hasUnpairedSurrogate } from "./surrogate.js";
import { signMessage } from "./sign.js";
import { signInkMessage, verifyInkSignature } from "./ink.js";

// A single backslash. Escape sequences are built by concatenation (bs + "uD800")
// so the test source never contains a literal escape a tool might fold into a
// real code point.
const bs = "\\";
const hi = bs + "uD83D";
const lo = bs + "uDE00";
const loneHi = bs + "uD800";
const loneLo = bs + "uDC00";

describe("containsLoneSurrogateEscape (raw JSON text)", () => {
  const accepts: Array<[string, string]> = [
    ["plain", `{"note":"hello"}`],
    ["empty string", `{"note":""}`],
    ["no strings", `{"n":1,"b":true,"x":null}`],
    ["valid pair", `{"note":"${hi}${lo}"}`],
    ["lowercase valid pair", `{"note":"${bs}ud83d${bs}ude00"}`],
    ["literal escaped backslash u", `{"note":"${bs}${bs}uD800"}`],
    ["bmp escape", `{"note":"${bs}u0041"}`],
    ["surrogate-shaped hex outside string", loneHi],
  ];
  for (const [name, raw] of accepts) {
    it(`accepts ${name}`, () => expect(containsLoneSurrogateEscape(raw)).toBe(false));
  }

  const rejects: Array<[string, string]> = [
    ["lone high", `{"note":"${loneHi}"}`],
    ["lone low", `{"note":"${loneLo}"}`],
    ["lowercase lone high", `{"note":"${bs}ud800"}`],
    ["high then bmp", `{"note":"${loneHi}${bs}u0041"}`],
    ["high split by char", `{"note":"${loneHi}x${loneLo}"}`],
    ["high then literal backslash u", `{"note":"${loneHi}${bs}${bs}uDC00"}`],
    ["lone high in key", `{"${loneHi}":"v"}`],
    ["lone low in array", `{"a":["x","${loneLo}"]}`],
    ["truncated lone high", `{"note":"${loneHi}`],
  ];
  for (const [name, raw] of rejects) {
    it(`rejects ${name}`, () => expect(containsLoneSurrogateEscape(raw)).toBe(true));
  }
});

describe("hasUnpairedSurrogate (parsed value)", () => {
  const H = String.fromCharCode(0xd800); // lone high
  const L = String.fromCharCode(0xdc00); // lone low
  const pair = String.fromCharCode(0xd83d, 0xde00); // valid astral pair

  it("accepts plain values", () => {
    expect(hasUnpairedSurrogate({ note: "hello", n: 1, ok: true })).toBe(false);
  });
  it("accepts a valid surrogate pair", () => {
    expect(hasUnpairedSurrogate({ note: pair })).toBe(false);
  });
  it("rejects a lone high in a value", () => {
    expect(hasUnpairedSurrogate({ note: H })).toBe(true);
  });
  it("rejects a lone low in a value", () => {
    expect(hasUnpairedSurrogate({ note: L })).toBe(true);
  });
  it("rejects a lone surrogate in a key", () => {
    expect(hasUnpairedSurrogate({ [H]: "v" })).toBe(true);
  });
  it("rejects a lone surrogate nested in an array", () => {
    expect(hasUnpairedSurrogate({ a: ["x", { b: L }] })).toBe(true);
  });
});

// The raw-text scanner (receiver boundary) and the parsed-object walk (signer
// boundary) must flag the same set: a lone surrogate detected on either side is
// detected on the other after JSON.parse / serialization.
describe("raw scan and object walk agree", () => {
  const bodies = [
    `{"note":"hello"}`,
    `{"note":"${hi}${lo}"}`,
    `{"note":"${loneHi}"}`,
    `{"note":"${loneLo}"}`,
    `{"a":["x","${loneLo}"]}`,
    `{"deep":{"k":"${loneHi}"}}`,
  ];
  for (const body of bodies) {
    it(`agree on ${body.slice(0, 24)}`, () => {
      const rawResult = containsLoneSurrogateEscape(body);
      const objResult = hasUnpairedSurrogate(JSON.parse(body));
      expect(objResult).toBe(rawResult);
    });
  }
});

describe("signMessage integration", () => {
  it("refuses to sign a body containing a lone surrogate", async () => {
    const key = new Uint8Array(32).fill(7);
    await expect(
      signMessage({ note: String.fromCharCode(0xd800) }, key),
    ).rejects.toThrow(/surrogate/);
  });
  it("signs an ordinary body", async () => {
    const key = new Uint8Array(32).fill(7);
    await expect(signMessage({ note: "hello" }, key)).resolves.toMatch(/^[A-Za-z0-9_-]{86}$/);
  });
});

describe("signInkMessage / verifyInkSignature integration", () => {
  const base = {
    method: "POST",
    path: "/ink/v1/x/intent",
    recipientDid: "tulpa:z",
    timestamp: "2026-06-11T00:00:00.000Z",
  };
  it("signInkMessage refuses a body with a lone surrogate", async () => {
    const key = new Uint8Array(32).fill(7);
    await expect(
      signInkMessage({ ...base, body: { note: String.fromCharCode(0xd800) } }, key),
    ).rejects.toThrow(/surrogate/);
  });
  it("verifyInkSignature rejects a body with a lone surrogate", async () => {
    const ok = await verifyInkSignature(
      { ...base, body: { note: String.fromCharCode(0xd800) } },
      "A".repeat(86),
      new Uint8Array(32),
    );
    expect(ok).toBe(false);
  });
});

File: go/ink/surrogate_test.go
package ink

import "testing"

// bs is a single backslash. Surrogate escape sequences are built by
// concatenation (bs + "uD800") so the test source never contains a literal
// \uXXXX escape that a tool might fold into a real code point.
const bs = "\\"

func TestContainsLoneSurrogateEscape(t *testing.T) {
	hi := bs + "uD83D" // high surrogate escape of an astral pair
	lo := bs + "uDE00" // matching low surrogate escape
	loneHi := bs + "uD800"
	loneHiUpperBound := bs + "uDBFF"
	loneLo := bs + "uDC00"
	loneLoUpperBound := bs + "uDFFF"
	lowerLoneHi := bs + "ud800"
	bmpA := bs + "u0041"
	litBackslashU := bs + bs + "uD800" // an escaped backslash then literal uD800

	cases := []struct {
		name string
		raw  string
		want bool
	}{
		// Accepted: no unpaired surrogate.
		{"plain string", `{"note":"hello"}`, false},
		{"empty string", `{"note":""}`, false},
		{"no strings", `{"n":1,"b":true,"x":null}`, false},
		{"valid surrogate pair escaped", `{"note":"` + hi + lo + `"}`, false},
		{"lowercase valid pair escaped", `{"note":"` + bs + "ud83d" + bs + "ude00" + `"}`, false},
		{"literal escaped backslash u", `{"note":"` + litBackslashU + `"}`, false},
		{"bmp unicode escape", `{"note":"` + bmpA + `"}`, false},
		{"raw utf8 multibyte", `{"note":"héllo 😀"}`, false},
		{"two valid pairs escaped", `{"a":"` + hi + lo + `","b":"` + hi + lo + `"}`, false},
		{"surrogate-shaped hex outside any string", loneHi, false},

		// Rejected: an unpaired surrogate inside a string.
		{"lone high at end", `{"note":"` + loneHi + `"}`, true},
		{"lone low", `{"note":"` + loneLo + `"}`, true},
		{"lowercase lone high", `{"note":"` + lowerLoneHi + `"}`, true},
		{"high then bmp escape", `{"note":"` + loneHi + bmpA + `"}`, true},
		{"high split by char", `{"note":"` + loneHi + "x" + loneLo + `"}`, true},
		{"high then literal escaped backslash u", `{"note":"` + loneHi + bs + bs + "uDC00" + `"}`, true},
		{"lone high in object key", `{"` + loneHi + `":"v"}`, true},
		{"lone low in array element", `{"a":["x","` + loneLo + `"]}`, true},
		{"truncated lone high at input end", `{"note":"` + loneHi, true},
		{"high boundary DBFF lone", `{"note":"` + loneHiUpperBound + `"}`, true},
		{"low boundary DFFF lone", `{"note":"` + loneLoUpperBound + `"}`, true},
	}
	for _, c := range cases {
		if got := ContainsLoneSurrogateEscape([]byte(c.raw)); got != c.want {
			t.Errorf("%s: ContainsLoneSurrogateEscape(%q) = %v, want %v", c.name, c.raw, got, c.want)
		}
	}
}

func TestParseSignedBody(t *testing.T) {
	if _, err := ParseSignedBody([]byte(`{"note":"hello"}`)); err != nil {
		t.Errorf("valid body rejected: %v", err)
	}
	if _, err := ParseSignedBody([]byte(`{"note":"` + bs + `uD800"}`)); err == nil {
		t.Errorf("lone surrogate body accepted")
	}
	if _, err := ParseSignedBody([]byte(`{not json`)); err == nil {
		t.Errorf("malformed JSON accepted")
	}
}

func TestParseHex4(t *testing.T) {
	cases := []struct {
		in    string
		want  int
		valid bool
	}{
		{"D800", 0xD800, true},
		{"d800", 0xD800, true},
		{"0041", 0x0041, true},
		{"FFFF", 0xFFFF, true},
		{"00", 0, false},   // too short
		{"D80G", 0, false}, // non-hex digit
		{"xyza", 0, false},
	}
	for _, c := range cases {
		got, ok := parseHex4([]byte(c.in), 0)
		if ok != c.valid || (ok && got != c.want) {
			t.Errorf("parseHex4(%q) = (%d,%v), want (%d,%v)", c.in, got, ok, c.want, c.valid)
		}
	}
}
